---
title: "analysis_code"
author: "Diana_Gutierrez"
date: "8/25/2020"
output: html_document
---
####EXAMPLE CODE########

This markdown contains examples of code you can use for your analysis. There is no preset way of analyzing your data and there are decisions that need to be made depending on your data quality, characteristics and other experimental factors so you should use this code and apply it to what you need to evaluate in your especific analysis

###Exploratory code 

```{r}
hist(sample_sums(ps), main="Histogram: Read Counts", xlab="Total Reads", 
     border="blue", col="green", las=1, breaks=12)
```



###Aglomerating data 

Here we are agglomerating the counts to the Phylum-level and then renaming the ASVs to make them more descriptive. We can see that we have 10 Phyla. The ASV information (i.e. refseq and taxonomy for one of the ASVs in each Phylum) gets carried along for the ride (we can typically ignore this or you can remove these components if you prefer).


```{r}
(ps_phylum <- tax_glom(ps, "Phylum"))
taxa_names(ps_phylum)
taxa_names(ps_phylum) <- tax_table(ps_phylum)[, 2]
taxa_names(ps_phylum)
otu_table(ps_phylum)[1:5, c(1:3, 5, 7)]

```

###Subset Data

```{r}
#We can also subset taxa…
(ps_bacteroides <- subset_taxa(ps, Genus == "Bacteroides"))
tax_table(ps_bacteroides)
prune_taxa(taxa_sums(ps) > 100, ps) 
filter_taxa(ps, function(x) sum(x > 10) > (0.1*length(x)), TRUE)   

```
With the above commands we can quickly see that we have:
A total of 3 ASVs classified as Bacteroides
A total of 99 ASVs seen at least 100 times across all samples
A total of 135 taxa seen at least 10 times in at least 10% of samples
This highlights how we might use phyloseq as a tool to filter taxa prior to statistical analysis.



*###Graphic code###* 
###Alpha-Diversity 

```{r}
head(estimate_richness(ps2))
```





```{r}

(p <- plot_richness(ps2, x = "diet_group", color = "sex", measures = c("Observed", "Chao1", "Shannon")))
(ps <- plot_richness(ps, x = "diet_group", color = "sex", measures = c("Observed", "Chao1", "Shannon")))

p + labs(x = "", y = "\nAlpha Diversity Measure\n") + 
  theme_bw()



```
###Beta diversity ordination 

```{r}
ps_bray <- ordinate(ps2, "NMDS", "bray")
plot_ordination(ps, ps_bray, type="samples", color="diet_group", shape = "sex") + geom_point(size = 3) 

```




###TAXA AGGLOMERATION 

Because  agglomerate taxa at the Genus level (combine all with the same name) and remove all taxa without genus level assignment

```{r}
#We can do it using a taxonomic method 

# How many genera would be present after filtering?
length(get_taxa_unique(ps2, taxonomic.rank = "Genus"))
## [1] 49
ps3 = tax_glom(ps2.prev, "Genus", NArm = TRUE)

sum(colSums(otu_table(ps3)))
```

```{r}
#Or by determining a tree hight at which you want to conglomerate the measures 

h1 = 0.4
ps4 = tip_glom(ps2.prev, h = h1)
sum(colSums(otu_table(ps4)))

```

We can generate a tree picture from each method and compare it to the previous tree "Before Agglomeration"

```{r}
multiPlotTitleTextSize = 15
p2tree = plot_tree(ps2, method = "treeonly",
                   ladderize = "left",
                   title = "Before Agglomeration") +
  theme(plot.title = element_text(size = multiPlotTitleTextSize))
p3tree = plot_tree(ps3, method = "treeonly",
                   ladderize = "left", title = "By Genus") +
  theme(plot.title = element_text(size = multiPlotTitleTextSize))
p4tree = plot_tree(ps4, method = "treeonly",
                   ladderize = "left", title = "By Height") +
  theme(plot.title = element_text(size = multiPlotTitleTextSize))
# group plots together
grid.arrange(nrow = 1, p2tree, p3tree, p4tree)
```

It is usually necessary to transform microbiome count data to account for differences in library size, variance, scale, etc. The phyloseq package provides a flexible interface for defining new functions to accomplish these transformations of the abundance values via the function transform_sample_counts(). The first argument to this function is the phyloseq object you want to transform, and the second argument is an R function that defines the transformation. The R function is applied sample-wise, expecting that the first unnamed argument is a vector of taxa counts in the same order as the phyloseq object. Additional arguments are passed on to the function specified in the second argument, providing an explicit means to include pre-computed values, previously defined parameters/thresholds, or any other object that might be appropriate for computing the transformed values of interest.

This example begins by defining a custom plot function, plot_abundance(), that uses phyloseq’s function to define a relative abundance graphic. We will use this to compare more easily differences in scale and distribution of the abundance values in our phyloseq object before and after transformation.
```{r}
plot_abundance = function(physeq,title = "",
                          Facet = "Order", Color = "sex"){
  # Arbitrary subset, based on Phylum, for plotting
  p1f = subset_taxa(ps2.prev, Phylum %in% c("Firmicutes"))
  mphyseq = psmelt(p1f)
  mphyseq <- subset(mphyseq, Abundance > 0)
  ggplot(data = mphyseq, mapping = aes_string(x = "sex",y = "Abundance",
                              color = Color, fill = Color)) +
    geom_violin(fill = NA) +
    geom_point(size = 1, alpha = 0.3,
               position = position_jitter(width = 0.3)) +
    facet_wrap(facets = Facet) + scale_y_log10()+
    theme(legend.position="none")
}
```
The transformation in this case converts the counts from each sample into their frequencies, often referred to as proportions or relative abundances. This function is so simple that it is easiest to define it within the function call to transform_sample_counts().
```{r}
# Transform to relative abundance. Save as new object.
ps2ra = transform_sample_counts(ps2.prev, function(x){x / sum(x)})
```
Now we plot the abundance values before and after transformation.

```{r}
plotBefore = plot_abundance(ps2,"")
plotAfter = plot_abundance(ps2ra,"")
# Combine each plot into one graphic.
grid.arrange(nrow = 2,  plotBefore, plotAfter)
```
```{r}
psOrd = subset_taxa(ps2ra, Order == "Bacillales")
plot_abundance(psOrd, Facet = "Family", Color = NULL)
```

Do some simple ordination looking for outlier samples, first we variance stabilize the data with a log transform, the perform PCoA using bray’s distances

```{r}
logt  = transform_sample_counts(ps3, function(x) log(1 + x) )
out.pcoa.logt <- ordinate(logt, method = "PCoA", distance = "Bray")
evals <- out.pcoa.logt$values$Eigenvalues
plot_ordination(logt, out.pcoa.logt, type = "samples", 
                color = "sex", shape = "diet_group") + labs(col = "Diet Group") +
  coord_fixed(sqrt(evals[2] / evals[1]))


```

To generate a heatmaap 

```{r}
 plot_heatmap(ps3, method = "NMDS", distance = "bray")
```
its too cluttered so we should only count the most abundant AVSs

```{r}
 total = median(sample_sums(ps3))
  ps3_abund <- filter_taxa(ps3, function(x) sum(x > total*0.20) > 0, TRUE)
  ps3_abund
  
    plot_heatmap(ps3_abund, method = "NMDS", distance = "bray")

```




```{r}
.cran_packages <- c( "shiny","miniUI", "caret", "pls", "e1071", "ggplot2", "randomForest", "dplyr", "ggrepel", "nlme", "devtools",
                  "reshape2", "PMA", "structSSI", "ade4",
                  "ggnetwork", "intergraph", "scales")
.github_packages <- c("jfukuyama/phyloseqGraphTest")
.bioc_packages <- c("genefilter", "impute")
# Install CRAN packages (if not already installed)
.inst <- .cran_packages %in% installed.packages()
if (any(!.inst)){
  install.packages(.cran_packages[!.inst],repos = "http://cran.rstudio.com/")
}
.inst <- .github_packages %in% installed.packages()
if (any(!.inst)){
  devtools::install_github(.github_packages[!.inst])
}

.inst <- .bioc_packages %in% installed.packages()
if(any(!.inst)){
  source("http://bioconductor.org/biocLite.R")
  biocLite(.bioc_packages[!.inst])
}
```





